### PR TITLE
fix(ts): make template types consistent

### PR DESCRIPTION
**Pull request by shortcuts**
**Summary**

Follow up of https://github.com/algolia/instantsearch.js/pull/4774

In some cases, we were exporting our Templates types as required for our widgets, while they are in reality optional and partial.

To avoid our users to provide extra logic over our types and to keep it consistent with our usage, the types are now:
- Exported as optional in the widget.
- Export as required in the component.

**Result**

The exported types are now correct.

**Ticket**

[DX-1778](https://algolia.atlassian.net/browse/DX-1778)

### CODE REVIEW

File: src/components/RefinementList/RefinementList.tsx [hunk 1]
```diff
@@ -42,6 +47,12 @@ export type RefinementListComponentCSSClasses = RefinementListRequired & {
     Pick<HierarchicalMenuComponentCSSClasses, 'childList' | 'parentItem'>
   >;
 
+// Templates types
```
Comment (francoischalifour) [inline src/components/RefinementList/RefinementList.tsx:50]: Not sure these comments are useful.

File: src/components/RefinementList/RefinementList.tsx [hunk 2]
```diff
@@ -313,7 +322,7 @@ class RefinementList<TTemplates extends Templates> extends Component<
           placeholder={this.props.searchPlaceholder}
           disabled={shouldDisableSearchBox}
           cssClasses={this.props.cssClasses.searchable!}
-          templates={templates}
+          templates={{ ...defaultSearchBoxTemplates, ...templates }}
```
Comment (francoischalifour) [inline src/components/RefinementList/RefinementList.tsx:325]: That would avoid having line 316?

```suggestion
          templates={{ ...defaultSearchBoxTemplates, ...this.props.searchBoxTemplateProps.templates }}
```

File: src/widgets/answers/answers.tsx [hunk 3]
```diff
@@ -53,20 +53,20 @@ export type AnswersTemplates = {
   /**
    * Template to use for the header. This template will receive an object containing `hits` and `isLoading`.
    */
-  header: Template<{
+  header?: Template<{
```
Comment (francoischalifour) [inline src/widgets/answers/answers.tsx:56]: We can `AnswersTemplates = Partial<{/* ... */}>` like elsewhere? (it's less error prone)

File: src/widgets/geo-search/geo-search.ts [hunk 4]
```diff
@@ -23,13 +23,13 @@ const suit = component('GeoSearch');
 
 export type GeoSearchTemplates = {
   /** Template to use for the marker. */
-  HTMLMarker: Template<GeoHit>;
+  HTMLMarker?: Template<GeoHit>;
```
Comment (francoischalifour) [inline src/widgets/geo-search/geo-search.ts:26]: Same comment about `Partial`.

File: src/widgets/refinement-list/refinement-list.tsx [hunk 5]
```diff
@@ -20,41 +21,42 @@ import { Template, WidgetFactory, RendererOptions } from '../../types';
...
+import { SearchBoxComponentTemplates } from '../../components/SearchBox/SearchBox';
 
 const withUsage = createDocumentationMessageGenerator({
   name: 'refinement-list',
 });
 const suit = component('RefinementList');
 const searchBoxSuit = component('SearchBox');
 
-type RefinementListOwnTemplates = {
+export type RefinementListOwnTemplates = {
```
Comment (francoischalifour) [inline src/widgets/refinement-list/refinement-list.tsx:32]: If we export it, we can rename it to that for consistency IMO:

```suggestion
export type RefinementListTemplates = {
```

File: src/widgets/refinement-list/refinement-list.tsx [hunk 6]
```diff
@@ -190,14 +192,14 @@ export type RefinementListWidgetParams = {
...
-  templates?: Partial<RefinementListTemplates>;
+  templates?: RefinementListTemplates;
   /**
    * CSS classes to add to the wrapping elements.
    */
   cssClasses?: RefinementListCSSClasses;
 };
 
-export const defaultTemplates: RefinementListOwnTemplates = {
+export const defaultTemplates: RefinementListWidgetTemplates = {
```
Comment (francoischalifour) [inline src/widgets/refinement-list/refinement-list.tsx:202]: We don't use the `WidgetTemplates` suffix elsewhere.

File: src/widgets/refinement-list/refinement-list.tsx [hunk 7]
```diff
@@ -413,15 +419,21 @@ const refinementList: RefinementListWidget = function refinementList(
...
+    templates,
     searchBoxTemplates: {
-      submit: userTemplates.searchableSubmit,
-      reset: userTemplates.searchableReset,
-      loadingIndicator: userTemplates.searchableLoadingIndicator,
+      submit: templates.searchableSubmit || searchBoxDefaultTemplates.submit,
+      reset: templates.searchableReset || searchBoxDefaultTemplates.reset,
+      loadingIndicator:
+        templates.searchableLoadingIndicator ||
+        searchBoxDefaultTemplates.loadingIndicator,
```
Comment (francoischalifour) [inline src/widgets/refinement-list/refinement-list.tsx:436]: Why did this logic have to change? IIUC it slightly changes the behavior.

File: src/widgets/stats/stats.tsx [hunk 8]
```diff
@@ -139,7 +139,7 @@ const stats: StatsWidget = widgetParams => {
   const {
     container,
     cssClasses: userCssClasses = {} as StatsCSSClasses,
-    templates = defaultTemplates,
+    templates: userTemplates = {} as StatsTemplates,
```
Comment (francoischalifour) [inline src/widgets/stats/stats.tsx:142]: In some other widget we were able to remove the `as StatsTemplates` cast. Why not here?

File: src/widgets/refinement-list/refinement-list.tsx (hunk 5 above)
Comment (shortcuts) [reply src/widgets/refinement-list/refinement-list.tsx:32]: I've also had this concern with the `CSSClasses` types but it only refer to the `refinement-list` widget itself, without the `searchable` templates so that's why I've decided to keep this naming.

[`RefinementListTemplates`](https://github.com/algolia/instantsearch.js/blob/74c7158c98a6f0d058e113687d072f570fedc518/src/widgets/refinement-list/refinement-list.tsx#L62-L63) is the other existing type for both the `OwnTemplates` & the searchable ones

File: src/widgets/stats/stats.tsx (hunk 8 above)
Comment (shortcuts) [reply src/widgets/stats/stats.tsx]: Good catch, this is indeed not needed: https://github.com/algolia/instantsearch.js/pull/4785/commits/e95210c1a0568ad92493815760bc6dc9a8464f0b

File: src/components/RefinementList/RefinementList.tsx (hunk 1 above)
Comment (shortcuts) [reply src/components/RefinementList/RefinementList.tsx]: True, removed in: https://github.com/algolia/instantsearch.js/pull/4785/commits/97c8c9a82afbb7a7462a57fee4a25a44580ff14b

File: src/widgets/refinement-list/refinement-list.tsx (hunk 5 above)
Comment (francoischalifour) [reply src/widgets/refinement-list/refinement-list.tsx:32]: We only want to export the complete type then, not this non-complete one?

File: src/widgets/refinement-list/refinement-list.tsx (hunk 7 above)
Comment (shortcuts) [reply src/widgets/refinement-list/refinement-list.tsx:436]: If I've understood correctly how it behaves, the logic changed because the renderer is supposed to receive all the templates. It was previously not throwing an error since we partially accepted templates, but we were spreading the `searchBoxDefaultTemplates` anyway.

Thinking again, since we spread the `searchBoxDefaultTemplates` anyway, it probably make sense to make it partial in the `renderer`, wdyt?

File: src/widgets/refinement-list/refinement-list.tsx (hunk 6 above)
Comment (shortcuts) [reply src/widgets/refinement-list/refinement-list.tsx:203]: Since it's only internal, does it make sense to you to have keep the naming with this implementation?https://github.com/algolia/instantsearch.js/pull/4785/commits/83c614ed93f39f4be32ad2ed7721b7de5b69ad56

File: src/widgets/voice-search/voice-search.tsx [hunk 9]
```diff
@@ -113,16 +110,26 @@ const voiceSearch: VoiceSearch = widgetParams => {
     button: cx(suit({ descendantName: 'button' }), userCssClasses.button),
     status: cx(suit({ descendantName: 'status' }), userCssClasses.status),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/voice-search/voice-search.tsx:116]: isn't this code already done by prepareTemplateProps? maybe we should pass templates, defaultTemplates to the renderer, and let prepareTemplateProps deal with it. Does that give the same outcome?

Comment (shortcuts) [reply src/widgets/voice-search/voice-search.tsx:116]: This indeed gives the same outcome, but some widgets don't yet use `prepareTemplateProps`, or/and don't have typed `renderer` params. I think it would be best to handle all of this in a PR where we type the `renderer` params and use `prepareTemplateProps` at the same time. 

wdyt? 

wdyt?

File: src/widgets/answers/answers.tsx (hunk 3 above)
Comment (Haroenv) [reply src/widgets/answers/answers.tsx:56]: yes, I agree, we should probably have all templates as non-optional in the template type and use Partial in the parameter (nothing in the component needed as change then)

Comment (shortcuts) [reply src/widgets/answers/answers.tsx:56]: Since it's unanimous, I'll do a stacked PR that converts all types to `export type T = Partial<{ ... }>`  once this one is finished

File: src/components/RefinementList/RefinementList.tsx [hunk 10]
```diff
@@ -42,6 +47,12 @@ export type RefinementListComponentCSSClasses = RefinementListRequired & {
     Pick<HierarchicalMenuComponentCSSClasses, 'childList' | 'parentItem'>
   >;
 
+// Templates types
+
+export type RefinementListWidgetTemplates = Required<
+  RefinementListOwnTemplates
```
[the remark this answers is not in the document]
Comment (shortcuts) [reply src/components/RefinementList/RefinementList.tsx]: No, the templates (`Own` and `Searchables`) are grouped in the widget under the same `templates` key, but in the Component, they both have their own key (`templates`, `searchBoxTemplates`).

File: src/components/RefinementList/RefinementList.tsx [hunk 11]
```diff
@@ -42,6 +47,12 @@ export type RefinementListComponentCSSClasses = RefinementListRequired & {
     Pick<HierarchicalMenuComponentCSSClasses, 'childList' | 'parentItem'>
   >;
 
+// Templates types
+
+export type RefinementListWidgetTemplates = Required<
```
[the remark this answers is not in the document]
Comment (shortcuts) [reply src/components/RefinementList/RefinementList.tsx]: Yep! Done in https://github.com/algolia/instantsearch.js/pull/4785/commits/e731f890f197e56576e017f47a721b60e30c54b5

File: src/components/Panel/Panel.tsx [hunk 12]
```diff
@@ -16,13 +16,17 @@ export type PanelComponentCSSClasses = ComponentCSSClasses<
   Omit<PanelCSSClasses, 'collapseIcon'>
 >;
 
+export type PanelComponentTemplates<
+  TWidget extends UnknownWidgetFactory
+> = Required<PanelTemplates<TWidget>>;
```
[the remark this answers is not in the document]
Comment (shortcuts) [reply src/components/Panel/Panel.tsx:21]: Yes, they are in the `PanelTemplates` type so we just require them

File: src/widgets/refinement-list/refinement-list.tsx (hunk 5 above)
Comment (shortcuts) [reply src/widgets/refinement-list/refinement-list.tsx]: It's now exported to be used in the `Component` and create the `RefinementListComponentTemplates` type, do you think it shouldn't be exposed at all? (https://github.com/algolia/instantsearch.js/pull/4785/commits/e731f890f197e56576e017f47a721b60e30c54b5)

In that case, I could create the `Component` type directly from the widget 🤔

File: src/components/Panel/Panel.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (Haroenv) [reply src/components/Panel/Panel.tsx:21]: misread this originally as the templates for TWidget

File: src/components/RefinementList/RefinementList.tsx [hunk 13]
```diff
@@ -302,18 +311,17 @@ class RefinementList<TTemplates extends Templates> extends Component<
...
     const searchBox = this.props.searchFacetValues && (
       <div className={this.props.cssClasses.searchBox}>
         <SearchBox
           ref={this.searchBox}
           placeholder={this.props.searchPlaceholder}
           disabled={shouldDisableSearchBox}
           cssClasses={this.props.cssClasses.searchable!}
-          templates={templates}
+          templates={{
+            ...defaultSearchBoxTemplates,
```
Comment (Haroenv) [inline src/components/RefinementList/RefinementList.tsx:322]: I'm not sure, but changing the defaults could be a breaking change, what are the templates exactly?

Don't they get defaulted again in the searchbox component? 

Thanks for checking!

File: src/components/RefinementList/__tests__/RefinementList-test.tsx [hunk 14]
```diff
@@ -606,6 +604,7 @@ describe('RefinementList', () => {
       const templates = {
         item: (item: RefinementListItemData) => item.value,
         showMoreText: '',
+        searchableNoResults: 'No results',
```
Comment (Haroenv) [inline src/components/RefinementList/__tests__/RefinementList-test.tsx:607]: why this change?

File: src/widgets/answers/answers.tsx (hunk 3 above)
Comment (Haroenv) [reply src/widgets/answers/answers.tsx:56]: could you do the stacked PR before we merge this one? That way there's less churn on master

File: src/widgets/menu-select/menu-select.tsx [hunk 15]
```diff
@@ -92,7 +95,9 @@ const renderer = ({
   isFirstRendering: boolean
 ) => {
   if (isFirstRendering) {
-    renderState.templateProps = prepareTemplateProps({
+    renderState.templateProps = prepareTemplateProps<
+      MenuSelectComponentTemplates
```
Comment (Haroenv) [inline src/widgets/menu-select/menu-select.tsx:99]: does this not get inferred from the arguments?

File: src/widgets/refinement-list/refinement-list.tsx [hunk 16]
```diff
@@ -161,6 +100,71 @@ type RefinementListSearchableCSSClasses = {
 export type RefinementListCSSClasses = RefinementListOwnCSSClasses &
   RefinementListSearchableCSSClasses;
 
+// Templates types
```
Comment (Haroenv) [inline src/widgets/refinement-list/refinement-list.tsx:103]:
```suggestion
```

If the grouping becomes too complex, we can move it into its own file, but I don't think that's needed for now

File: src/widgets/refinement-list/refinement-list.tsx (hunk 7 above)
Comment (Haroenv) [reply src/widgets/refinement-list/refinement-list.tsx:438]: agreed, since we spread in the template, we should accept partial searchbox templates

File: src/widgets/sort-by/sort-by.tsx [hunk 17]
```diff
@@ -23,15 +23,15 @@ export type SortByWidgetCssClasses = {
   /**
    * CSS classes added to the outer `<div>`.
    */
-  root: string | string[];
+  root?: string | string[];
```
Comment (Haroenv) [inline src/widgets/sort-by/sort-by.tsx:26]: better in a separate PR, but fine for now I think

File: src/widgets/toggle-refinement/toggle-refinement.tsx [hunk 18]
```diff
@@ -48,7 +49,9 @@ const renderer = ({
   isFirstRendering: boolean
 ) => {
   if (isFirstRendering) {
-    renderState.templateProps = prepareTemplateProps({
+    renderState.templateProps = prepareTemplateProps<
+      ToggleRefinementComponentTemplates
+    >({
       defaultTemplates,
```
Comment (Haroenv) [inline src/widgets/toggle-refinement/toggle-refinement.tsx:55]: in some of the others you spread the defaultTemplates, but not here. Which pattern do you think is better? I'd like it to be the same in all widgets I think

**Review by Haroenv — COMMENTED** [review]
overall almost good to merge, although the Partial<{...}> should probably be stacked + deciding what to do with spreading defaultTemplates or equaling to defaultTemplates

File: src/widgets/answers/answers.tsx (hunk 3 above)
Comment (shortcuts) [reply src/widgets/answers/answers.tsx:56]: I'll do a stacked PR on https://github.com/algolia/instantsearch.js/pull/4791 right after!

Comment (Haroenv) [reply src/widgets/answers/answers.tsx:56]: thought about it again, and since it touches the css types too, let's do a completely separate PR, no need to worry about master churn in this case :)

File: src/widgets/breadcrumb/breadcrumb.tsx [hunk 19]
```diff
@@ -153,6 +159,10 @@ const breadcrumb: BreadcrumbWidget = function breadcrumb(widgetParams) {
     ),
     link: cx(suit({ descendantName: 'link' }), userCssClasses.link),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/breadcrumb/breadcrumb.tsx:165]: this isn't needed normally, right?

File: src/widgets/clear-refinements/clear-refinements.tsx [hunk 20]
```diff
@@ -116,6 +116,10 @@ const clearRefinements: ClearRefinementsWidget = widgetParams => {
       userCssClasses.disabledButton
     ),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/clear-refinements/clear-refinements.tsx:122]: this also isn't needed, as it gets defaulted in prepareTemplateProps

File: src/widgets/hierarchical-menu/hierarchical-menu.tsx [hunk 21]
```diff
@@ -335,6 +339,10 @@ const hierarchicalMenu: HierarchicalMenuWidget = function hierarchicalMenu(
       userCssClasses.disabledShowMore
     ),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/hierarchical-menu/hierarchical-menu.tsx:345]: not needed

File: src/widgets/menu-select/menu-select.tsx (hunk 15 above)
Comment (shortcuts) [reply src/widgets/menu-select/menu-select.tsx]: It now does since we typed the `defaultTemplates`: https://github.com/algolia/instantsearch.js/pull/4785/commits/8f538d06fbe57af7d8ff0f6b73cecd88bb90997b

File: src/widgets/hits/__tests__/defaultTemplates-test.ts [hunk 22]
```diff
@@ -20,6 +20,7 @@ describe('hits defaultTemplates', () => {
   }
 }`;
 
+    // @ts-expect-error `item` is considered as not callable
     expect(defaultTemplates.item(item)).toBe(expected);
```
Comment (Haroenv) [inline src/widgets/hits/__tests__/defaultTemplates-test.ts:24]:
```suggestion
    expect(
      typeof defaultTemplates.item === 'function' && defaultTemplates.item(item)
    ).toBe(expected);
```

File: src/widgets/hits/hits.tsx [hunk 23]
```diff
@@ -150,6 +150,10 @@ const hits: HitsWidget = function hits(widgetParams) {
     list: cx(suit({ descendantName: 'list' }), userCssClasses.list),
     item: cx(suit({ descendantName: 'item' }), userCssClasses.item),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/hits/hits.tsx:156]: idem

File: src/widgets/infinite-hits/__tests__/defaultTemplates-test.ts [hunk 24]
```diff
@@ -20,6 +20,7 @@ describe('hits defaultTemplates', () => {
   }
 }`;
 
+    // @ts-expect-error `item` is considered as not callable
     expect(defaultTemplates.item(item)).toBe(expected);
```
Comment (Haroenv) [inline src/widgets/infinite-hits/__tests__/defaultTemplates-test.ts:24]:
```suggestion
    expect(
      typeof defaultTemplates.item === 'function' && defaultTemplates.item(item)
    ).toBe(expected);
```

File: src/widgets/infinite-hits/infinite-hits.tsx [hunk 25]
```diff
@@ -218,6 +218,10 @@ const infiniteHits: InfiniteHitsWidget = widgetParams => {
       userCssClasses.disabledLoadMore
     ),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/infinite-hits/infinite-hits.tsx:224]: idem

File: src/widgets/menu-select/menu-select.tsx [hunk 26]
```diff
@@ -142,6 +145,10 @@ const menuSelect: MenuSelectWidget = function menuSelect(widgetParams) {
     select: cx(suit({ descendantName: 'select' }), userCssClasses.select),
     option: cx(suit({ descendantName: 'option' }), userCssClasses.option),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/menu-select/menu-select.tsx:151]: idem

File: src/widgets/menu/menu.tsx [hunk 27]
```diff
@@ -206,6 +210,10 @@ const menu: MenuWidget = function menu(widgetParams) {
       userCssClasses.disabledShowMore
     ),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/menu/menu.tsx:216]: same

File: src/widgets/numeric-menu/numeric-menu.tsx [hunk 28]
```diff
@@ -195,6 +197,10 @@ const numericMenu: NumericMenuWidget = function numericMenu(widgetParams) {
       userCssClasses.labelText
     ),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/numeric-menu/numeric-menu.tsx:203]: idem

File: src/widgets/rating-menu/rating-menu.tsx [hunk 29]
```diff
@@ -238,6 +248,10 @@ const ratingMenu: RatingMenuWidget = function ratingMenu(widgetParams) {
     label: cx(suit({ descendantName: 'label' }), userCssClasses.label),
     count: cx(suit({ descendantName: 'count' }), userCssClasses.count),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/rating-menu/rating-menu.tsx:254]: here

File: src/widgets/search-box/defaultTemplates.ts [hunk 30]
```diff
@@ -1,6 +1,7 @@
-/* eslint max-len: 0 */
+import { SearchBoxComponentTemplates } from '../../components/SearchBox/SearchBox';
 
-export default {
+/* eslint max-len: 0 */
```
Comment (Haroenv) [inline src/widgets/search-box/defaultTemplates.ts:3]: still needed?

File: src/widgets/toggle-refinement/toggle-refinement.tsx [hunk 31]
```diff
@@ -153,6 +156,10 @@ const toggleRefinement = function toggleRefinement(widgetParams) {
       userCssClasses.labelText
     ),
   };
+  const templates = {
+    ...defaultTemplates,
+    ...userTemplates,
+  };
```
Comment (Haroenv) [inline src/widgets/toggle-refinement/toggle-refinement.tsx:162]: idem

File: src/widgets/voice-search/voice-search.tsx (hunk 9 above)
Comment (Haroenv) [reply src/widgets/voice-search/voice-search.tsx:116]: and still here

**Review by Haroenv — COMMENTED** [review]
all right, this is a great effort! let's still do a global search for `userTemplates` and avoid it wherever the templates get prepared after :)

File: src/widgets/voice-search/voice-search.tsx (hunk 9 above)
Comment (shortcuts) [reply src/widgets/voice-search/voice-search.tsx:116]: This one doesn't have the `prepareTemplateProps` so I've kept the spread here

File: src/widgets/toggle-refinement/toggle-refinement.tsx (hunk 31 above)
Comment (shortcuts) [reply src/widgets/toggle-refinement/toggle-refinement.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/search-box/defaultTemplates.ts (hunk 30 above)
Comment (shortcuts) [reply src/widgets/search-box/defaultTemplates.ts]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/rating-menu/rating-menu.tsx (hunk 29 above)
Comment (shortcuts) [reply src/widgets/rating-menu/rating-menu.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/numeric-menu/numeric-menu.tsx (hunk 28 above)
Comment (shortcuts) [reply src/widgets/numeric-menu/numeric-menu.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/menu/menu.tsx (hunk 27 above)
Comment (shortcuts) [reply src/widgets/menu/menu.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/menu-select/menu-select.tsx (hunk 26 above)
Comment (shortcuts) [reply src/widgets/menu-select/menu-select.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/infinite-hits/infinite-hits.tsx (hunk 25 above)
Comment (shortcuts) [reply src/widgets/infinite-hits/infinite-hits.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/infinite-hits/__tests__/defaultTemplates-test.ts (hunk 24 above)
Comment (shortcuts) [reply src/widgets/infinite-hits/__tests__/defaultTemplates-test.ts]: A bit different but done here: https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a#diff-ccf1a9e06ea635549b22aba227e2ed2825fe1731acbd6b6196b12df424fd8158R7

File: src/widgets/hits/__tests__/defaultTemplates-test.ts (hunk 22 above)
Comment (shortcuts) [reply src/widgets/hits/__tests__/defaultTemplates-test.ts]: A bit different but done here: https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a#diff-ccf1a9e06ea635549b22aba227e2ed2825fe1731acbd6b6196b12df424fd8158R7

File: src/widgets/hierarchical-menu/hierarchical-menu.tsx (hunk 21 above)
Comment (shortcuts) [reply src/widgets/hierarchical-menu/hierarchical-menu.tsx]: https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/hits/hits.tsx (hunk 23 above)
Comment (shortcuts) [reply src/widgets/hits/hits.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/clear-refinements/clear-refinements.tsx (hunk 20 above)
Comment (shortcuts) [reply src/widgets/clear-refinements/clear-refinements.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/breadcrumb/breadcrumb.tsx (hunk 19 above)
Comment (shortcuts) [reply src/widgets/breadcrumb/breadcrumb.tsx]: see https://github.com/algolia/instantsearch.js/pull/4785/commits/61ea462ec076c7fd7a38a33f8d0bde9344d72d0a

File: src/widgets/query-rule-custom-data/query-rule-custom-data.tsx [hunk 32]
```diff
@@ -52,9 +57,11 @@ const renderer = ({
   containerNode: HTMLElement;
   cssClasses: QueryRuleCustomDataComponentCSSClasses;
   renderState: {
-    templateProps?: PreparedTemplateProps<QueryRuleCustomDataTemplates>;
+    templateProps?: PreparedTemplateProps<
+      QueryRuleCustomDataComponentTemplates
+    >;
```
Comment (Haroenv) [inline src/widgets/query-rule-custom-data/query-rule-custom-data.tsx:62]: renderState is defined, but we don't use it, and rely on spreading in the widget itself?

**Review by Haroenv — APPROVED** [review]
perfect!
